generator/index.ts

```diff
@@ -0,0 +1,86 @@
+import * as path from 'path';
+
+import {FileData, RootCards, SerializedGuideJson} from './file-types.js';
+import * as fs from './fsp.js';
+import {readLearningPath, readTopLevelFile} from './readers.js';
+import {writeLearningPath, writeRootCards, writeSerializedGuideJson, writeTopLevelFile} from './writers.js';
+
+const TOP_LEVEL_DIRECTORY = path.resolve(__dirname, '..', '..');
+const BUILD_OUTPUT = path.resolve(TOP_LEVEL_DIRECTORY, 'build_output', 'en');
+const CONTENT = path.resolve(TOP_LEVEL_DIRECTORY, 'content');
+
+async function readContentDirectory(): Promise<FileData[]> {
+  const contentFilesOrDirectories =
+      await fs.readdir(CONTENT, {withFileTypes: true});
+
+  if (typeof contentFilesOrDirectories[0] === 'string') {
+    throw new Error(`You need at least Node 10.10.0 for this script`);
+  }
+
+  return Promise.all(contentFilesOrDirectories.map(async (fileOrDirectory) => {
+    const resolvedName = path.resolve(CONTENT, fileOrDirectory.name);
+
+    if (fileOrDirectory.isDirectory()) {
+      // Don't try to parse the images folder as a learning path
+      if (await fs.exists(path.resolve(resolvedName, 'guides.yaml'))) {
+        return readLearningPath(resolvedName, fileOrDirectory.name);
+      }
+
+      return {name: fileOrDirectory.name};
+    }
+
+    return readTopLevelFile(resolvedName, fileOrDirectory.name);
+  }));
+}
+
+function shouldPreserveFileForBuildOutput(_file: FileData) {
+  return true;
+}
+
+async function writeBuildOutput(files: FileData[]) {
+  await fs.rimraf(BUILD_OUTPUT);
+  await fs.mkdirp(BUILD_OUTPUT);
+
+  const rootCards: RootCards = {paths: []};
+  const allGuides: SerializedGuideJson[] = [];
+
+  for (const file of files) {
+    if ('body' in file) {
+      await writeTopLevelFile(BUILD_OUTPUT, file);
+    } else if ('topics' in file) {
+      await writeLearningPath(BUILD_OUTPUT, file);
+      rootCards.paths.push({...file, href: `/${file.name}`});
```

Comment:
I was wondering if this section and the block below that builds `allGuides` should be handled by readers as well. It feels odd for `index.ts` to own this logic.

Reply:
Sounds like a plan.